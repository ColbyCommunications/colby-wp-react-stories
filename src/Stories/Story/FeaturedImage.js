import PropTypes from 'prop-types';
import styled from 'styled-components';

const StyledImage = styled.img`
  width: 100%;
  height: auto;
`;

const FeaturedImage = ({ altText, featuredImage }) => {
  try {
    const data =
      featuredImage.media_details.sizes.large ||
      featuredImage.media_details.sizes.medium;

    return (
      <StyledImage
        alt={altText}
        src={data.source_url}
        width={data.width}
        height={data.height}
      />
    );
  } catch (e) {
    return null;
  }
};

FeaturedImage.propTypes = {
  altText: PropTypes.string.isRequired,
  featuredImage: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default FeaturedImage;
