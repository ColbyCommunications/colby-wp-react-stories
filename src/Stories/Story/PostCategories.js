import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { withStateContext, withActionContext } from '../Context';

const StyledContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  padding: 1rem;
`;

const StyledButton = styled.button`
  flex: 0 0 46%;
  padding: 0.375rem;
  margin-right: 0.25rem;
  margin-bottom: 0.5rem;
  margin-left: 0.25rem;
  font-size: 66.6%;
  line-height: 1.2;
  color: black;
  color: var(--color, black);
  text-transform: uppercase;
  cursor: pointer;
  background: rgba(0, 0, 0, 0.075);
  background: var(--light-gray, rgba(0, 0, 0, 0.075));
  border: 1px solid rgba(0, 0, 0, 0.15);
  border-radius: 2px;
  transition: 0.2s background-color;

  &:hover {
    background: rgba(0, 0, 0, 0.1);
    background: var(--background-hover-color, rgba(0, 0, 0, 0.1));
  }
`;

export const PostCategories = ({
  post, setActiveCategory, categories,
}) => (
  <StyledContainer>
    {categories
      .filter(category => post.categories.indexOf(category.id) !== -1)
      .map(({ id, name, slug }) => (
        <StyledButton
          className={`Stories__category--${slug}`}
          key={id}
          onClick={() => {
            setActiveCategory(String(id));
          }}
          dangerouslySetInnerHTML={{ __html: name }}
        />
        ))}
  </StyledContainer>
);

PostCategories.propTypes = {
  post: PropTypes.objectOf(PropTypes.any).isRequired,
  setActiveCategory: PropTypes.func.isRequired,
  categories: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default withStateContext(withActionContext(PostCategories));

