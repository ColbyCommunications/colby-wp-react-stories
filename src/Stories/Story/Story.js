/* eslint react/no-danger: 0 */
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import distanceInWords from 'date-fns/distance_in_words';

import Categories from './PostCategories';
import FeaturedImage from './FeaturedImage';

import { withActionContext } from '../Context';

const StyledStory = styled.article`
  position: relative;
  display: flex;
  flex-direction: column;
  min-width: 0;
  margin-bottom: 1rem;
  font-size: 89.1%;
  word-wrap: break-word;
  background-color: #fff;
  background-clip: border-box;
  border: 1px solid rgba(0, 0, 0, 0.125);
  border-radius: 2px;

  time {
    font-size: 89.1%;
  }
`;

const StyledStoryBody = styled.div`
  flex: 1 1 auto;
  padding: 0.75rem 1.25rem;
`;

const StyledStoryTitle = styled.h1`
  font-size: 119%;
  line-height: 1.333;
`;

class Story extends React.Component {
  static propTypes = {
    fetchPostMedia: PropTypes.func.isRequired,
    post: PropTypes.objectOf(PropTypes.any).isRequired,
    setActiveCategory: PropTypes.func.isRequired,
  }

  componentDidMount() {
    if (this.props.post.featured_media !== 0) {
      this.props.fetchPostMedia(this.props.post.ID, this.props.post.featuredMedia);
    }
  }

  render = ({ post, setActiveCategory } = this.props) => (
    <StyledStory key={post.id}>
      {post.featuredImage && <FeaturedImage
        altText={post.title.rendered}
        featuredImage={post.featuredImage}
      />}
      <StyledStoryBody>
        <StyledStoryTitle>
          <a
            href={post.link}
            dangerouslySetInnerHTML={{ __html: post.title.rendered }}
          />
        </StyledStoryTitle>
        <time dateTime={post.date}>
          {distanceInWords(new Date(), post.date, {
            addSuffix: true,
          })}
        </time>
        <p dangerouslySetInnerHTML={{ __html: post.excerpt.rendered }} />
      </StyledStoryBody>
      <Categories post={post} setActiveCategory={setActiveCategory} />
    </StyledStory>
  );
}

export default withActionContext(Story);
