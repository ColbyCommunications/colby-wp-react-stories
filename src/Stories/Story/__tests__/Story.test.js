import React from 'react';
import renderer from 'react-test-renderer';

import Story from '../Story';
import { posts, media } from '../../__tests__/testData';

describe('Story', () => {
  it('renders', () => {
    const tree = renderer.create(<Story
      post={posts[0]}
      fetchPostMedia={() => null}
      setActiveCategory={() => null}
    />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders with a featured image', () => {
    const post = { ...posts[0], featuredImage: media };
    const tree = renderer.create(<Story
      post={post}
      fetchPostMedia={() => null}
      setActiveCategory={() => null}
    />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders with featured_media property', () => {
    const tree = renderer.create(<Story
      post={posts[1]}
      fetchPostMedia={() => null}
      setActiveCategory={() => null}
    />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
