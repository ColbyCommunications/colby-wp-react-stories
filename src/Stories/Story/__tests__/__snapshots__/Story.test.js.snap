// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story renders 1`] = `
<article
  className="sc-bxivhb bAWjzE"
>
  <div
    className="sc-ifAKCX deqhXU"
  >
    <h1
      className="sc-EHOje wDHIL"
    >
      <a
        dangerouslySetInnerHTML={
          Object {
            "__html": "Hello",
          }
        }
        href={undefined}
      />
    </h1>
    <time
      dateTime={undefined}
    >
      almost NaN years ago
    </time>
    <p
      dangerouslySetInnerHTML={
        Object {
          "__html": "Hello, this is the excerpt.",
        }
      }
    />
  </div>
  <div
    className="sc-bdVaJa fEpVJS"
  />
</article>
`;

exports[`Story renders with a featured image 1`] = `
<article
  className="sc-bxivhb bAWjzE"
>
  <img
    alt="Hello"
    className="sc-htpNat kmMOsy"
    height={100}
    src="http://my-image"
    width={200}
  />
  <div
    className="sc-ifAKCX deqhXU"
  >
    <h1
      className="sc-EHOje wDHIL"
    >
      <a
        dangerouslySetInnerHTML={
          Object {
            "__html": "Hello",
          }
        }
        href={undefined}
      />
    </h1>
    <time
      dateTime={undefined}
    >
      almost NaN years ago
    </time>
    <p
      dangerouslySetInnerHTML={
        Object {
          "__html": "Hello, this is the excerpt.",
        }
      }
    />
  </div>
  <div
    className="sc-bdVaJa fEpVJS"
  />
</article>
`;

exports[`Story renders with featured_media property 1`] = `
<article
  className="sc-bxivhb bAWjzE"
>
  <div
    className="sc-ifAKCX deqhXU"
  >
    <h1
      className="sc-EHOje wDHIL"
    >
      <a
        dangerouslySetInnerHTML={
          Object {
            "__html": "My cool post",
          }
        }
        href={undefined}
      />
    </h1>
    <time
      dateTime={undefined}
    >
      almost NaN years ago
    </time>
    <p
      dangerouslySetInnerHTML={
        Object {
          "__html": "Hello, this is a cool post",
        }
      }
    />
  </div>
  <div
    className="sc-bdVaJa fEpVJS"
  />
</article>
`;
